Post-#112261 pointer += integer diagnostics suggest replacing with unused value

### Code

```Rust
fn main() {
	let ga_z = [];
	let mut known_rgbs_groups = 0 as *mut u32;
	known_rgbs_groups += ga_z.len();
}
```

### Current output

```Shell
stable:

error[E0368]: binary assignment operation `+=` cannot be applied to type `*mut u32`
 --> test.rs:4:2
  |
4 |     known_rgbs_groups += ga_z.len();
  |     -----------------^^^^^^^^^^^^^^
  |     |
  |     cannot use `+=` on type `*mut u32`
  |
help: consider using `wrapping_add` or `add` for pointer + {integer}
  |
4 |     known_rgbs_groups.wrapping_add(ga_z.len());
  |                      ~~~~~~~~~~~~~~          +

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0368`.

nightly:
error[E0368]: binary assignment operation `+=` cannot be applied to type `*mut u32`
 --> test.rs:4:2
  |
4 |     known_rgbs_groups += ga_z.len();
  |     -----------------^^^^^^^^^^^^^^
  |     |
  |     cannot use `+=` on type `*mut u32`
  |
help: consider using `wrapping_add` or `add` for pointer + {integer}
  |
4 -     known_rgbs_groups += ga_z.len();
4 +     known_rgbs_groups.wrapping_add(ga_z.len());
  |

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0368`.
```

### Desired output

```Shell
known_rgbs_groups = known_rgbs_groups.add(ga_z.len());
```

### Rationale and extra context

`known_rgbs_groups.add(ga_z.len());` doesn't do anything 💀 

### Other cases

```Rust

```

### Rust Version

```Shell
rustc 1.84.1 (e71f9a9a9 2025-01-27)
binary: rustc
commit-hash: e71f9a9a98b0faf423844bf0ba7438f29dc27d58
commit-date: 2025-01-27
host: x86_64-pc-windows-gnu
release: 1.84.1
LLVM version: 19.1.5


rustc 1.87.0-nightly (827a0d638 2025-02-18)
binary: rustc
commit-hash: 827a0d638dabc9a22c56f9c37a557568f86ac76c
commit-date: 2025-02-18
host: x86_64-pc-windows-gnu
release: 1.87.0-nightly
LLVM version: 20.1.0
```

### Anything else?

This is #112261, and in 20 comments and 4 review rounds no-one tried += lol

--- a/compiler/rustc_hir_typeck/src/op.rs
+++ b/compiler/rustc_hir_typeck/src/op.rs
@@ -234,7 +234,6 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         // us do better coercions than we would be able to do otherwise,
         // particularly for things like `String + &String`.
         let rhs_ty_var = self.next_ty_var(rhs_expr.span);
-
         let result = self.lookup_op_method(
             (lhs_expr, lhs_ty),
             Some((rhs_expr, rhs_ty_var)),
@@ -689,6 +688,57 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                                             ".offset_from(".to_owned(),
                                         ),
                                         (rhs_expr.span.shrink_to_hi(), ") }".to_owned()),
+                                    ],
+                                    Applicability::MaybeIncorrect,
+                                );
+                            }
+                        }
+                        _ => {}
+                    }
+                }
+
+                let lhs_name_str = match lhs_expr.kind {
+                    hir::ExprKind::Path(hir::QPath::Resolved(_, path)) => {
+                        path.segments.last().map_or("_".to_string(), |s| s.ident.to_string())
+                    }
+                    _ => self
+                        .tcx
+                        .sess
+                        .source_map()
+                        .span_to_snippet(lhs_expr.span)
+                        .unwrap_or("_".to_string()),
+                };
+
+                if op.span().can_be_used_for_suggestions() {
+                    match op {
+                        Op::AssignOp(Spanned { node: hir::AssignOpKind::AddAssign, .. })
+                            if lhs_ty.is_raw_ptr() && rhs_ty.is_integral() =>
+                        {
+                            err.multipart_suggestion(
+                                "consider using `add` or `wrapping_add` to do pointer arithmetic",
+                                vec![
+                                    (lhs_expr.span.shrink_to_lo(), format!("{} = ", lhs_name_str)),
+                                    (
+                                        lhs_expr.span.between(rhs_expr.span),
+                                        ".wrapping_add(".to_owned(),
+                                    ),
+                                    (rhs_expr.span.shrink_to_hi(), ")".to_owned()),
+                                ],
+                                Applicability::MaybeIncorrect,
+                            );
+                        }
+                        Op::AssignOp(Spanned { node: hir::AssignOpKind::SubAssign, .. }) => {
+                            if lhs_ty.is_raw_ptr() && rhs_ty.is_integral() {
+                                err.multipart_suggestion(
+                                    "consider using `sub` or `wrapping_sub` to do pointer arithmetic",
+                                    vec![
+                                        (lhs_expr.span.shrink_to_lo(), format!("{} = ", lhs_name_str)),
+                                        (
+                                            lhs_expr.span.between(rhs_expr.span),
+                                            ".wrapping_sub(".to_owned(),
+
+                                        ),
+                                        (rhs_expr.span.shrink_to_hi(), ")".to_owned()),
                                     ],
                                     Applicability::MaybeIncorrect,
                                 );

--- a/tests/ui/typeck/pointer-arith-assign.fixed
+++ b/tests/ui/typeck/pointer-arith-assign.fixed
@@ -1,0 +1,20 @@
+//@ run-rustfix
+#![allow(dead_code)]
+#![allow(unused_variables)]
+#![allow(unused_assignments)]
+
+fn test_add_assign_raw_pointer() {
+    let mut arr = [0u8; 10];
+    let mut _ptr = arr.as_mut_ptr();
+
+    _ptr = _ptr.wrapping_add(2); //~ ERROR binary assignment operation `+=` cannot be applied to type `*mut u8` [E0368]
+}
+
+fn test_sub_assign_raw_pointer() {
+    let mut arr = [0u8; 10];
+    let mut _ptr = arr.as_mut_ptr();
+
+    _ptr = _ptr.wrapping_sub(2); //~ ERROR binary assignment operation `-=` cannot be applied to type `*mut u8` [E0368]
+}
+
+fn main() {}

--- a/tests/ui/typeck/pointer-arith-assign.rs
+++ b/tests/ui/typeck/pointer-arith-assign.rs
@@ -1,0 +1,20 @@
+//@ run-rustfix
+#![allow(dead_code)]
+#![allow(unused_variables)]
+#![allow(unused_assignments)]
+
+fn test_add_assign_raw_pointer() {
+    let mut arr = [0u8; 10];
+    let mut _ptr = arr.as_mut_ptr();
+
+    _ptr += 2; //~ ERROR binary assignment operation `+=` cannot be applied to type `*mut u8` [E0368]
+}
+
+fn test_sub_assign_raw_pointer() {
+    let mut arr = [0u8; 10];
+    let mut _ptr = arr.as_mut_ptr();
+
+    _ptr -= 2; //~ ERROR binary assignment operation `-=` cannot be applied to type `*mut u8` [E0368]
+}
+
+fn main() {}

--- a/tests/ui/typeck/pointer-arith-assign.stderr
+++ b/tests/ui/typeck/pointer-arith-assign.stderr
@@ -1,0 +1,31 @@
+error[E0368]: binary assignment operation `+=` cannot be applied to type `*mut u8`
+  --> $DIR/pointer-arith-assign.rs:10:5
+   |
+LL |     _ptr += 2;
+   |     ----^^^^^
+   |     |
+   |     cannot use `+=` on type `*mut u8`
+   |
+help: consider using `add` or `wrapping_add` to do pointer arithmetic
+   |
+LL -     _ptr += 2;
+LL +     _ptr = _ptr.wrapping_add(2);
+   |
+
+error[E0368]: binary assignment operation `-=` cannot be applied to type `*mut u8`
+  --> $DIR/pointer-arith-assign.rs:17:5
+   |
+LL |     _ptr -= 2;
+   |     ----^^^^^
+   |     |
+   |     cannot use `-=` on type `*mut u8`
+   |
+help: consider using `sub` or `wrapping_sub` to do pointer arithmetic
+   |
+LL -     _ptr -= 2;
+LL +     _ptr = _ptr.wrapping_sub(2);
+   |
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0368`.

